File: server/src/users/store.ts
import { randomUUID } from 'node:crypto';

/**
 * Người dùng của V-Market.
 *
 * Điểm quan trọng: `role` và `sellerId` là DỮ LIỆU CỦA V-MARKET,
 * không đến từ V-App. V-App chỉ trả về danh tính (user_id) — nó không
 * có khái niệm buyer/seller. Việc ai là người bán là chuyện của V-Market.
 *
 * Ngày 1 lưu trong bộ nhớ. Ngày 2 chuyển sang DB thật cùng lúc làm
 * seed buyer / seller A / seller B.
 */

export type UserRole = 'BUYER' | 'SELLER';

export type MarketUser = {
  id: string;
  /** Khoá liên kết sang V-App. Duy nhất. */
  vappUserId: string;
  role: UserRole;
  /** Chỉ có khi role === 'SELLER'. MVP: một seller một shop. */
  sellerId: string | null;
  name: string | null;
  phoneNumber: string | null;
};

const byVAppUserId = new Map<string, MarketUser>();

/**
 * Vai trò định sẵn cho tài khoản demo. Đây là bảng của V-Market,
 * không phải dữ liệu lấy từ V-App — cố ý tách rời để thấy rõ ranh giới.
 */
const SEED_ROLES: Record<string, { role: UserRole; sellerId: string | null }> = {
  '11111111-1111-4111-8111-111111111111': { role: 'BUYER', sellerId: null },
  '22222222-2222-4222-8222-222222222222': {
    role: 'SELLER',
    sellerId: 'seller-a',
  },
  '33333333-3333-4333-8333-333333333333': {
    role: 'SELLER',
    sellerId: 'seller-b',
  },
};

export function findByVAppUserId(vappUserId: string): MarketUser | undefined {
  return byVAppUserId.get(vappUserId);
}

export function createUser(input: {
  vappUserId: string;
  name: string | null;
  phoneNumber: string | null;
}): MarketUser {
  const seeded = SEED_ROLES[input.vappUserId];

  const user: MarketUser = {
    id: randomUUID(),
    vappUserId: input.vappUserId,
    // Mặc định là BUYER: người dùng mới trên một sàn TMĐT là người mua.
    // Nâng lên SELLER là hành động riêng (tạo shop — ngày 3).
    role: seeded?.role ?? 'BUYER',
    sellerId: seeded?.sellerId ?? null,
    name: input.name,
    phoneNumber: input.phoneNumber,
  };

  byVAppUserId.set(user.vappUserId, user);
  return user;
}

/** Chỉ dùng trong test. */
export function resetUserStore(): void {
  byVAppUserId.clear();
}

File: server/app/users/store.py
"""V-Market users.

`role` is V-Market's data, not V-App's. V-App only returns an identity —
it has no notion of buyer or seller.
"""

import uuid
from typing import Literal

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base

UserRole = Literal["BUYER", "SELLER"]


class MarketUser(Base):
    __tablename__ = "users"

    id: Mapped[str] = mapped_column(primary_key=True)
    # Unique: two rows for one person would split their orders in half.
    vapp_user_id: Mapped[str] = mapped_column(unique=True, index=True)
    # Everyone arrives a BUYER. Opening a shop is what makes a SELLER, so
    # there is no list of privileged accounts anywhere.
    role: Mapped[str] = mapped_column(default="BUYER")
    name: Mapped[str | None] = mapped_column(default=None)
    phone_number: Mapped[str | None] = mapped_column(default=None)


async def find_by_id(session: AsyncSession, user_id: str) -> MarketUser | None:
    return await session.get(MarketUser, user_id)


async def find_by_vapp_user_id(
    session: AsyncSession, vapp_user_id: str
) -> MarketUser | None:
    return await session.scalar(
        select(MarketUser).where(MarketUser.vapp_user_id == vapp_user_id)
    )


async def create_user(
    session: AsyncSession,
    vapp_user_id: str,
    name: str | None,
    phone_number: str | None,
) -> MarketUser:
    user = MarketUser(
        id=str(uuid.uuid4()),
        vapp_user_id=vapp_user_id,
        role="BUYER",
        name=name,
        phone_number=phone_number,
    )
    session.add(user)
    await session.commit()
    await session.refresh(user)
    return user


async def promote_to_seller(
    session: AsyncSession, user: MarketUser
) -> MarketUser:
    """Called when a shop is opened. A seller keeps buying from other shops."""
    user.role = "SELLER"
    await session.commit()
    await session.refresh(user)
    return user
